odl/space/base_tensors.py

```diff
@@ -0,0 +1,741 @@
+﻿# Copyright 2014-2016 The ODL development group
+#
+# This file is part of ODL.
+#
+# ODL is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Base classes for implementations of tensor spaces."""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+
+import numpy as np
+
+from odl.set.sets import Set, RealNumbers, ComplexNumbers
+from odl.set.space import LinearSpace, LinearSpaceElement
+from odl.util import (
+    is_scalar_dtype, is_floating_dtype, is_real_dtype,
+    is_complex_floating_dtype, safe_int_conv,
+    arraynd_repr, arraynd_str, dtype_str, signature_string, indent_rows)
+from odl.util.ufuncs import TensorSpaceUfuncs
+from odl.util.utility import TYPE_MAP_R2C, TYPE_MAP_C2R
+
+
+__all__ = ('TensorSpace',)
+
+
+class TensorSpace(LinearSpace):
+
+    """Base class for sets of tensors of arbitrary data type.
+
+    A tensor is, in the most general sense, a multi-dimensional array
+    that allows operations per entry (keep the rank constant),
+    reductions / contractions (reduce the rank) and broadcasting
+    (raises the rank).
+    For non-numeric data type like ``object``, the range of valid
+    operations is rather limited since such a set of tensors does not
+    define a vector space.
+    Any numeric data type, on the other hand, is considered valid for
+    a tensor space, although certain operations - like division with
+    integer dtype - are not guaranteed to yield reasonable results.
+
+    Under these restrictions, all basic vector space operations are
+    supported by this class, along with reductions based on arithmetic
+    or comparison, and element-wise mathematical functions ("ufuncs").
+
+    See `this Wikipedia article <https://en.wikipedia.org/wiki/Tensor>`_
```

Comment:
I have a book I can cite. Perhaps there's also a more specific section at Wikipedia which is more in line with our case.